File: removeKfromList.java
Given a singly linked list of integers l and an integer k, remove all elements from list l that have a value equal to k.

Example

For l = [3, 1, 2, 3, 4, 5] and k = 3, the output should be
removeKFromList(l, k) = [1, 2, 4, 5];
For l = [1, 2, 3, 4, 5, 6, 7] and k = 10, the output should be
removeKFromList(l, k) = [1, 2, 3, 4, 5, 6, 7].

Solution:

// Singly-linked lists are already defined with this interface:
// class ListNode<T> {
//   ListNode(T x) {
//     value = x;
//   }
//   T value;
//   ListNode<T> next;
// }
//
ListNode<Integer> removeKFromList(ListNode<Integer> l, int k) {
    ListNode<Integer> current = l;
    ListNode<Integer> previous = l;
    while(current != null){
        if(current.value == k){
            if(current == l)
                l = l.next;
            else
                previous.next = current.next;
            current = current.next;
        }
        else{
            previous = current;
            current = current.next;
        }
    }
    return l;
}

File: Codesignal and Geekforgeeks Combined/removeKfromList.py
Given a singly linked list of integers l and an integer k, remove all elements from list l that have a value equal to k.

Example

For l = [3, 1, 2, 3, 4, 5] and k = 3, the output should be
removeKFromList(l, k) = [1, 2, 4, 5];
For l = [1, 2, 3, 4, 5, 6, 7] and k = 10, the output should be
removeKFromList(l, k) = [1, 2, 3, 4, 5, 6, 7].



Solution:

# Singly-linked lists are already defined with this interface:
# class ListNode(object):
#   def __init__(self, x):
#     self.value = x
#     self.next = None
#
def removeKFromList(l, k):
    current = l
    while current is not None:
        if current.value != k:
            break
        current = current.next
        
    new_head = current
    prev = None
    while current is not None:
        if current.value == k:
            if prev is not None:
                prev.next = current.next
            current = current.next
            continue
        prev = current
        current = current.next
        
    return new_head
